dozer-lambda/src/js/worker/mod.rs

```diff
@@ -0,0 +1,284 @@
+                field_names,
+                new_values,
+                old_values,
+            })
+            .await
+            .is_err()
+        {
+            error!("lambda runtime has panicked");
+            if let Some(handle) = self.handle.lock().await.take() {
+                if let Err(e) = handle.await {
+                    error!("lambda runtime panicked: {}", e);
+                }
+            }
+        }
+    }
+}
+
+/// Registration script should call `register_lambda`.
+async fn execute_registration_scripts(
+    main_worker: &mut MainWorker,
+    registration_scripts: Vec<String>,
+) -> Result<(), Error> {
+    for registration_script in registration_scripts {
+        let path = std::fs::canonicalize(registration_script.clone())
+            .map_err(|e| Error::CanonicalizePath(registration_script, e))?;
+        let module_specifier =
+            ModuleSpecifier::from_file_path(path).expect("we just canonicalized it");
+        info!("executing module {}", module_specifier);
+        main_worker.execute_side_module(&module_specifier).await?;
+    }
+    Ok(())
+}
+
+type LambdaHashMap = HashMap<NonZeroI32, v8::Global<v8::Function>>;
+
+#[op2(async)]
+fn register_lambda(
+    #[state] trigger: &Arc<Mutex<Trigger>>,
+    #[state] lambdas: &Rc<RefCell<LambdaHashMap>>,
+    #[string] endpoint: String,
+    lambda: &v8::Function,
+    #[global] lambda_global: v8::Global<v8::Function>,
+) -> impl Future<Output = Result<(), ReaderBuilderError>> {
+    register_lambda_impl(
+        trigger.clone(),
+        lambdas.clone(),
+        endpoint,
+        lambda.get_identity_hash(),
+        lambda_global,
+    )
+}
+
+async fn register_lambda_impl(
+    trigger: Arc<Mutex<Trigger>>,
+    lambdas: Rc<RefCell<LambdaHashMap>>,
+    endpoint: String,
+    identity: NonZeroI32,
+    lambda_global: v8::Global<v8::Function>,
+) -> Result<(), ReaderBuilderError> {
+    info!("registering lambda {} for endpoint {}", identity, endpoint);
+    trigger.lock().await.add_lambda(endpoint, identity).await?;
+    lambdas.borrow_mut().insert(identity, lambda_global);
+    Ok(())
+}
+
+extension!(
+    dozer_lambda_extension,
+    ops = [register_lambda],
+    options = {
+        trigger: Arc<Mutex<Trigger>>,
+        lambdas: Rc<RefCell<LambdaHashMap>>
+    },
+    state = |state, options| {
+        state.put(options.trigger);
+        state.put(options.lambdas);
+    },
+);
+
+#[derive(Debug)]
+enum Work {
+    CallLambda {
+        func: NonZeroI32,
+        operation_index: u64,
+        operation_type: &'static str,
+        field_names: Vec<String>,
+        new_values: Vec<Field>,
+        old_values: Option<Vec<Field>>,
+    },
+}
+
+async fn worker_loop(
+    mut runtime: JsRuntime,
+    mut work_receiver: Receiver<Work>,
+    lambdas: Rc<RefCell<LambdaHashMap>>,
+) {
+    loop {
+        match poll_fn(|cx| {
+            poll_work_and_event_loop(&mut runtime, &mut work_receiver, &lambdas.borrow(), cx)
```

Comment:
I think this might explode memory usage by calling the lambda function many times before it has completed. Should we instead run the lambda function to completion before calling it again?

Reply:
You're right. I'll add this to the list.